File: src/main/js/structure/ClassFile.js
function ClassFile(bytesArray) {

    var  bytes = bytesArray;

    var magicNumberSection = new MagicNumberSection(bytes);
    var minorBuildSection = new BuildSection(bytes, 4, "Minor build:");
    var majorBuildSection = new BuildSection(bytes, 6, "Major build:");
    var constantPoolCountSection = new CountSection(bytes, majorBuildSection.getEndByteIndex(), "Pool count");
    var constantPoolSection = new ConstantPoolSection(constantPoolCountSection.getCount(), bytes, 10);
    var accessFlagsSection = new AccessFlagsSection(bytes, constantPoolSection.getEndByteIndex());
    var thisClassSection = new ThisClassSection(bytes, accessFlagsSection.getEndByteIndex());
    var superClassSection = new SuperClassSection(bytes, thisClassSection.getEndByteIndex());
    var interfacesCountSection = new CountSection(bytes, superClassSection.getEndByteIndex(), "Interfaces count");
    var interfacesSection = new InterfacesSection(bytes, interfacesCountSection.getEndByteIndex(), interfacesCountSection.getCount());
    var fieldsCountSection = new CountSection(bytes, interfacesSection.getEndByteIndex(), "Fields count");
    var fieldsSection = new FieldsSection(bytes, fieldsCountSection.getEndByteIndex(), fieldsCountSection.getCount());
    var methodCountSection = new CountSection(bytes, fieldsSection.getEndByteIndex(), "Methods count");
    var methodsSection = new MethodsSection(bytes, methodCountSection.getEndByteIndex(), methodCountSection.getCount());

    this.length = function () {
        return bytes.length;
    }

    this.getByte = function(index) {
        return bytes[index];
    }

    this.getBytecode = function(index) {
        var section = this.getSection(index);
        if(section != null) {
            var bytecode = this.selectedIndex + section.getBytecode(index, constantPoolSection.getConstants());

            return bytecode;
        }
        else {
            return "Unmapped";
        }
    }

    this.getSelectedSection = function() {
        return this.selectedSection;
    }

    this.getSection = function(index) {
        this.selectedIndex = "";

        if(magicNumberSection.contains(index)) {
            this.selectedSection = magicNumberSection;
        }
        else if(minorBuildSection.contains(index)) {
            this.selectedSection = minorBuildSection;
        }
        else if(majorBuildSection.contains(index)) {
            this.selectedSection = majorBuildSection;
        }
        else if(constantPoolCountSection.contains(index)) {
            this.selectedSection = constantPoolCountSection;
        }
        else if(constantPoolSection.contains(index)) {
            this.selectedSection = constantPoolSection.getSection(index);

            var byteString = this.selectedSection.getTag();
            var formatted = formatAsOneByteHexString(byteString);

            this.selectedIndex = "(#" + (constantPoolSection.getSelectedIndex() + 1) + ") " + formatted + " ";
        }
        else if(accessFlagsSection.contains(index)) {
            this.selectedSection = accessFlagsSection;
        }
        else if(thisClassSection.contains(index)) {
            this.selectedSection = thisClassSection;
        }
        else if(superClassSection.contains(index)) {
            this.selectedSection = superClassSection;
        }
        else if(interfacesCountSection.contains(index)) {
            this.selectedSection = interfacesCountSection;
        }
        else if(interfacesSection.contains(index)) {
            this.selectedSection = interfacesSection;
        }
        else if(fieldsCountSection.contains(index)) {
            this.selectedSection = fieldsCountSection;
        }
        else if(fieldsSection.contains(index)) {
            this.selectedSection = fieldsSection.getFieldAtByteIndex(index);
        }
        else if(methodCountSection.contains(index)) {
            this.selectedSection = methodCountSection;
        }
        else if(methodsSection.contains(index)) {
            this.selectedSection = methodsSection.getMethodAtByteIndex(index);
        }
        else {
            this.selectedSection = null;
        }

        return this.selectedSection;
    }


    this.isConstantPoolIndex = function (index) {
        var rval = false;
        if(constantPoolSection) {
            rval = constantPoolSection.contains(index);
        }
        return rval;
    }

    this.getEndByteIndex = function (byteIndexStart) {
        if(this.selectedSection) {
            return this.selectedSection.getEndByteIndex(byteIndexStart);
        }

    }

    this.getInstructionSequence = function (hoverByteIndex) {
        // first determine the section, then handle appropriately
        return "";
    }

    this.getCssClass = function(section) {
        return section.getCssClass();
    }

    this.getSelectedRange = function() {
        var section = this.getSelectedSection();
        if(section != null) {
            return section.getStartByte() + "-" + section.length();
        }
        else {
            return "none";
        }
    }

    this.getConstants = function() {
        return constantPoolSection.getConstants();
    }
}

File: src/main/java/com/vgrazi/bytecodeexplorer/structure/ClassFile.java
package com.vgrazi.bytecodeexplorer.structure;


import com.vgrazi.bytecodeexplorer.structure.members.MethodSection;

/**
 * Created by vgrazi on 8/13/15.
 */
public class ClassFile {
    private final byte[] bytes;
    private final MagicNumberSection magicNumberSection;
    private final BuildSection majorBuildSection;
    private final BuildSection minorBuildSection;
    private final CountSection constantPoolCountSection;
    private final ConstantPoolSection constantPoolSection;
    private final ClassAccessFlagsSection classAccessFlagsSection;
    private final ClassSection thisClassSection;
    private final ClassSection superClassSection;
    private final CountSection interfaceCountSection;
    private final ClassSection[] interfaceSections;
    private final CountSection fieldCountSection;
    private final FieldSection[] fieldSections;
    private final CountSection methodCountSection;
    private final MethodSection[] methodSections;

    public ClassFile(byte[] bytes) {
        this.bytes = bytes;
        magicNumberSection = new MagicNumberSection(bytes);
        minorBuildSection = new BuildSection("Minor", bytes, 4);
        majorBuildSection = new BuildSection("Major", bytes, 6);
        constantPoolCountSection = new CountSection("Pool count", bytes, 8);
        constantPoolSection = new ConstantPoolSection(constantPoolCountSection.getCount(), bytes, 10);
        classAccessFlagsSection = new ClassAccessFlagsSection(bytes, constantPoolSection.getEndByteIndex());
        thisClassSection = new ClassSection("This class", bytes, classAccessFlagsSection.getEndByteIndex());
        superClassSection = new ClassSection("Super class", bytes, thisClassSection.getEndByteIndex());
        interfaceCountSection = new CountSection("Interface count", bytes, superClassSection.getEndByteIndex());
        interfaceSections = new ClassSection[interfaceCountSection.getCount()];
        for (int i = 0; i < interfaceSections.length; i++) {
            interfaceSections[i] = new ClassSection("Interface", bytes, interfaceCountSection.getEndByteIndex() + 2 * i);
        }
        fieldCountSection = new CountSection("Field count", bytes, interfaceCountSection.getEndByteIndex() + interfaceCountSection.getCount() * 2);
        fieldSections = new FieldSection[fieldCountSection.getCount()];
        int nextByte = fieldCountSection.getStartByteIndex() + fieldCountSection.length();
        for (int i = 0; i < fieldSections.length; i++) {
            FieldSection fieldSection = new FieldSection(bytes, nextByte);
            fieldSections[i] = fieldSection;
            nextByte += fieldSection.length();
        }
        methodCountSection = new CountSection("Method count", bytes, nextByte);
        methodSections = new MethodSection[methodCountSection.getCount()];
        nextByte = methodCountSection.getStartByteIndex() + methodCountSection.length();
        for (int i = 0; i < methodSections.length; i++) {
            MethodSection methodSection = new MethodSection(bytes, nextByte);
            methodSections[i] = methodSection;
            nextByte += methodSection.length();
        }
    }

    public byte[] getBytes() {
        return bytes;
    }

    public int length() {
        return bytes.length;
    }

    public byte getByte(int index) {
        return bytes[index];
    }

    public String getInstructionSequence(int hoverByteIndex) {
        // first determine the section, then handle appropriately
        ClassFileSection section = getSection(hoverByteIndex);
        if (section != null) {
            return section.toString();
        }
        return "unknown";
    }

    /**
     * Used mostly for color coding the sections of the hex table, the section number starts at 0 and sequences
     * for each section, so magic number is section 0, minor build section 1, major=section 2, then the sections within the constant
     * pool continue from there in sequence.
     *
     * @param byteIndex The index in the classfile of the byte we are checking
     */
    public int getSectionIndex(int byteIndex) {
        if (magicNumberSection.contains(byteIndex)) {
            return 0;
        } else if (minorBuildSection.contains(byteIndex)) {
            return 1;
        } else if (majorBuildSection.contains(byteIndex)) {
            return 2;
        } else if (constantPoolCountSection.contains(byteIndex)) {
            return 3;
        } else if (constantPoolSection.contains(byteIndex)) {
            return constantPoolSection.getSectionIndex(byteIndex) + 4;
        } else if (classAccessFlagsSection.contains(byteIndex)) {
            return constantPoolCountSection.getCount() + 3;
        } else if (thisClassSection.contains(byteIndex)) {
            return constantPoolCountSection.getCount() + 4;
        } else if (superClassSection.contains(byteIndex)) {
            return constantPoolCountSection.getCount() + 5;
        } else if (interfaceCountSection.contains(byteIndex)) {
            return constantPoolCountSection.getCount() + 6;
        }

        for (int i = 0; i < interfaceSections.length; i++) {

            if (interfaceSections[i].contains(byteIndex)) {
                int val = constantPoolCountSection.getCount() + 7 + i;
                return val;
            }
        }

        if (fieldCountSection.contains(byteIndex)) {
            int val = constantPoolCountSection.getCount() + 7 + interfaceSections.length;
            return val;
        }

        for (int i = 0; i < fieldSections.length; i++) {
            if (fieldSections[i].contains(byteIndex)) {
                int val = constantPoolCountSection.getCount() + interfaceCountSection.getCount() + 8 + i;
                return val;
            }
        }

        if(methodCountSection.contains(byteIndex)) {
            return constantPoolCountSection.getCount() + interfaceCountSection.getCount() + fieldCountSection.getCount() + 9;
        }

        for (int i = 0; i < methodSections.length; i++) {
            if (methodSections[i].contains(byteIndex)) {
                int val = constantPoolCountSection.getCount() + interfaceCountSection.getCount() + fieldCountSection.getCount() + 10 + i;
                return val;
            }
        }

        return constantPoolCountSection.getCount() + interfaceCountSection.getCount() + fieldSections.length + fieldCountSection.getCount() + 10;
    }

    /**
     * Used mostly for color coding the sections of the hex table, the section number starts at 0 and sequences
     * for each section, so magic number is section 0, minor build section 1, major=section 2, then the sections within the constant
     * pool continue from there in sequence.
     *
     * @param byteIndex The index in the classfile of the byte we are checking
     */

    public ClassFileSection getSection(int byteIndex) {
        if (magicNumberSection.contains(byteIndex)) {
            return magicNumberSection;
        } else if (minorBuildSection.contains(byteIndex)) {
            return minorBuildSection;
        } else if (majorBuildSection.contains(byteIndex)) {
            return majorBuildSection;
        } else if (constantPoolCountSection.contains(byteIndex)) {
            return constantPoolCountSection;
        } else if (classAccessFlagsSection.contains(byteIndex)) {
            return classAccessFlagsSection;
        } else if (thisClassSection.contains(byteIndex)) {
            return thisClassSection;
        } else if (superClassSection.contains(byteIndex)) {
            return superClassSection;
        } else if (interfaceCountSection.contains(byteIndex)) {
            return interfaceCountSection;
        } else if (fieldCountSection.contains(byteIndex)) {
            return fieldCountSection;
        }
        for (ClassSection interfacesSection : interfaceSections) {
            if (interfacesSection.contains(byteIndex)) {
                return interfacesSection;
            }
        }

        for (FieldSection fieldSection : fieldSections) {
            if (fieldSection.contains(byteIndex)) {
                return fieldSection;
            }
        }

        if(methodCountSection.contains(byteIndex)) {
            return methodCountSection;
        }

        for(MethodSection methodSection : methodSections) {
            if(methodSection.contains(byteIndex)) {
                return methodSection;
            }
        }

        int sectionIndex = constantPoolSection.getSectionIndex(byteIndex);
        return constantPoolSection.getSection(sectionIndex);
    }

}
